File: tc-sbt-runner-agent/src/main/java/jetbrains/buildServer/sbt/SbtRunnerBuildService.java
package jetbrains.buildServer.sbt;

import com.intellij.openapi.util.SystemInfo;
import com.intellij.openapi.util.text.StringUtil;
import jetbrains.buildServer.RunBuildException;
import jetbrains.buildServer.agent.AgentRuntimeProperties;
import jetbrains.buildServer.agent.runner.*;
import jetbrains.buildServer.messages.ErrorData;
import jetbrains.buildServer.runner.JavaRunnerConstants;
import jetbrains.buildServer.util.FileUtil;
import org.apache.commons.lang3.RandomStringUtils;
import org.apache.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.jar.JarFile;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@SuppressWarnings("WeakerAccess")
public class SbtRunnerBuildService extends BuildServiceAdapter {

    private static final Logger LOG = Logger.getLogger(SbtRunnerBuildServiceFactory.class.getName());

    private static final String SBT_LAUNCHER_JAR_NAME = "sbt-launch.jar";

    private static final String SBT_PATCH_JAR_NAME = "sbt-teamcity-logger.jar";

    private static final String SBT_PATCH_FOLDER_NAME = "tc_plugin";

    private static final String SBT_1_0_PATCH_FOLDER_NAME = "1.0";

    private static final String SBT_DISTRIB = "sbt-distrib";

    private static final String SBT_AUTO_HOME_FOLDER = "agent-sbt";

    private static final String INFILE_COMMANDS_FORMATTER = ";%s ;%s %s";

    private static final String RUN_INFILE_COMMANDS_FORMATTER = "< %s";

    private static final String SBT_INSTALLATION_STEP_NAME = "SBT installation";

    private static final String SBT_TEAMCITY_LOGGER_INSTALLATION = "SBT TeamCity logger installation";

    private static final String PATH = "PATH";

    private final static String[] SBT_JARS = new String[]{
            SBT_LAUNCHER_JAR_NAME,
            "classes"
    };
    public static final String BUILD_ACTIVITY_TYPE = "BUILD_ACTIVITY_TYPE";

    public static final Pattern LINES_TO_EXCLUDE = Pattern.compile("^\\[(error|warn)\\]",
            Pattern.CASE_INSENSITIVE + Pattern.MULTILINE);
    public static final Pattern KNOWN_SECTION_MESSAGE = Pattern.compile("^(##teamcity\\[compilationStarted|testSuiteStarted|testStarted)",
            Pattern.CASE_INSENSITIVE + Pattern.MULTILINE);
    private static final String SBT_PATCH_CLASS_NAME = "jetbrains.buildServer.sbtlogger.SbtTeamCityLogger";


    private final IvyCacheProvider myIvyCacheProvider;

    private final List<File> myFilesToDelete = new ArrayList<>();

    public enum SBTVersion {
        SBT_1_x,
        SBT_0_13_x
    }

    public SbtRunnerBuildService(IvyCacheProvider ivyCacheProvider) {
        myIvyCacheProvider = ivyCacheProvider;
    }

    @NotNull
    @Override
    public List<ProcessListener> getListeners() {
        return Collections.<ProcessListener>singletonList(new ProcessListenerAdapter() {
            boolean knownSectionStarts = false;

            @Override
            public void onStandardOutput(@NotNull String line) {
                if (!knownSectionStarts) {
                    //we need this otherwise we can hide important messages appeared before our own logger was started to print server messages
                    Matcher matcher = KNOWN_SECTION_MESSAGE.matcher(line);
                    if (matcher.find()) {
                        knownSectionStarts = true;
                    }
                }
                Matcher matcher = LINES_TO_EXCLUDE.matcher(line);
                if (matcher.find() && knownSectionStarts) {
                    //we don't want to duplicate lines
                    //sbt-tc-logger wraps WARN and ERROR messages
                    //we can exclude those messages from normal output
                    return;
                }
                getLogger().message(line);
            }

            public void onErrorOutput(@NotNull String line) {
                getLogger().warning(line);
            }

            @Override
            public void processFinished(int exitCode) {
                super.processFinished(exitCode);
            }
        });
    }


    @NotNull
    @Override
    public ProgramCommandLine makeProgramCommandLine() throws RunBuildException {

        List<String> jvmArgs = JavaRunnerUtil.extractJvmArgs(getRunnerParameters());

        String mainClassName = isAutoInstallMode() ? installSbt() : getMainClassName();

        SBTVersion sbtVersion = SbtVersionDetector.discoverSbtVersion(getWorkingDirectory(), getSbtLauncher(), jvmArgs, getLogger());

        copySbtTcLogger(sbtVersion);

        String javaHome = getJavaHome();
        String sbtHome = getSbtHome();
        getLogger().message("Java home set to: " + javaHome);
        getLogger().message("SBT home set to: " + sbtHome);

        JavaCommandLineBuilder cliBuilder = new JavaCommandLineBuilder();
        cliBuilder.setJavaHome(javaHome);
        cliBuilder.setBaseDir(getCheckoutDirectory().getAbsolutePath());

        cliBuilder.setSystemProperties(getVMProperties());
        Map<String, String> envVars = new HashMap<String, String>(getEnvironmentVariables());

        envVars.put(SbtRunnerConstants.SBT_HOME, sbtHome);
        envVars.put(JavaRunnerConstants.JAVA_HOME, javaHome);

        String path = envVars.get(PATH);
        envVars.put(PATH, javaHome + (!StringUtil.isEmpty(path) ? File.pathSeparator + path : ""));

        fixSbtSocketLengthIssueIfNeeded(envVars);

        cliBuilder.setEnvVariables(envVars);


        cliBuilder.setJvmArgs(jvmArgs);
        cliBuilder.setClassPath(getClasspath());
        cliBuilder.setMainClass(mainClassName);

        cliBuilder.setProgramArgs(getCommands(sbtVersion));

        cliBuilder.setWorkingDir(getWorkingDirectory().getAbsolutePath());

        return buildCommandline(cliBuilder);
    }

    private String getApplyCommand(SBTVersion sbtVersion) {
        String pathToPlugin = new File(getAutoInstallSbtFolder() + File.separator + getPatchFolder(sbtVersion) + File.separator + SBT_PATCH_JAR_NAME).getAbsolutePath();
        return "apply -cp \"" + pathToPlugin.replace('\\', '/') + "\" " + SBT_PATCH_CLASS_NAME;
    }

    private String getCheckStatusCommand() {
        return "sbt-teamcity-logger";
    }

    private String getJavaHome() throws RunBuildException {
        String javaHome = JavaRunnerUtil.findJavaHome(getRunnerParameters().get(JavaRunnerConstants.TARGET_JDK_HOME),
                getBuildParameters().getAllParameters(),
                AgentRuntimeProperties.getCheckoutDir(getRunnerParameters()));
        if (javaHome == null) throw new RunBuildException("Unable to find Java home");
        javaHome = FileUtil.getCanonicalFile(new File(javaHome)).getPath();
        return javaHome;
    }

    @NotNull
    private String getAutoInstallSbtFolder() {
        return getAgentTempDirectory() + File.separator + SBT_AUTO_HOME_FOLDER;
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    private void copyResources(String sourcePathInJar, String sourceName, File destinationDir) {
        destinationDir.mkdirs();
        File destination = new File(destinationDir, sourceName);
        FileUtil.copyResource(this.getClass(), sourcePathInJar + sourceName, destination);
        boolean done = destination.exists() && destination.isFile();
        if (done) {
            getLogger().message("Resource was copied to: " + destination);
        }
    }

    private String installSbt() {

        try {
            getLogger().activityStarted(SBT_INSTALLATION_STEP_NAME, "'Auto' mode was selected in SBT runner plugin settings", BUILD_ACTIVITY_TYPE);
            getLogger().message("SBT will be install to: " + getAutoInstallSbtFolder());
            copyResources("/" + SBT_DISTRIB + "/", SBT_LAUNCHER_JAR_NAME, new File(getAutoInstallSbtFolder() + File.separator + "bin"));
            getLogger().message("SBT home set to: " + getSbtHome());
            return getMainClassName();
        } catch (Exception e) {
            getLogger().internalError(ErrorData.PREPARATION_FAILURE_TYPE, "An error occurred during SBT installation", e);
            throw new IllegalStateException(e);
        } finally {
            getLogger().activityFinished(SBT_INSTALLATION_STEP_NAME, BUILD_ACTIVITY_TYPE);
        }

    }

    private void copySbtTcLogger(SBTVersion sbtVersion) {
        try {
            getLogger().activityStarted(SBT_TEAMCITY_LOGGER_INSTALLATION, BUILD_ACTIVITY_TYPE);
            String to = getAutoInstallSbtFolder() + File.separator + getPatchFolder(sbtVersion);
            String from = "/" + SBT_DISTRIB + "/"
                    + (sbtVersion == SBTVersion.SBT_1_x ? SBT_1_0_PATCH_FOLDER_NAME + "/" : "");
            getLogger().message(String.format("SBT logger %s will be installed from %s to %s", SBT_PATCH_JAR_NAME, from, to));
            copyResources(from, SBT_PATCH_JAR_NAME, new File(to));
        } catch (Exception e) {
            getLogger().internalError(ErrorData.PREPARATION_FAILURE_TYPE, "An error occurred during SBT installation", e);
            throw new IllegalStateException(e);
        } finally {
            getLogger().activityFinished(SBT_TEAMCITY_LOGGER_INSTALLATION, BUILD_ACTIVITY_TYPE);
        }

    }

    @NotNull
    private String getPatchFolder(SBTVersion sbtVersion) {
        return sbtVersion == SBTVersion.SBT_1_x ?
                SBT_PATCH_FOLDER_NAME + File.separator + SBT_1_0_PATCH_FOLDER_NAME : SBT_PATCH_FOLDER_NAME;
    }

    @NotNull
    private String getMainClassName() throws RunBuildException {
        try {
            File sbtLauncher = getSbtLauncher();
            getLogger().message("SBT main class name will be retrieved from: " + sbtLauncher);
            JarFile jf = new JarFile(sbtLauncher);
            return jf.getManifest().getMainAttributes().getValue("Main-Class");
        } catch (IOException e) {
            throw new RunBuildException("An error occurred during reading manifest in SBT launcher", e);
        }
    }

    @NotNull
    private File getSbtLauncher() {
        String sbtHome = getSbtHome();
        File jarDir = new File(sbtHome, "bin");
        return new File(jarDir, SBT_LAUNCHER_JAR_NAME);
    }


    @NotNull
    private ProgramCommandLine buildCommandline(@NotNull final JavaCommandLineBuilder cliBuilder) throws
            RunBuildException {
        try {
            return cliBuilder.build();
        } catch (CannotBuildCommandLineException e) {
            throw new RunBuildException(e.getMessage());
        }
    }

    @NotNull
    private Map<String, String> getVMProperties() throws RunBuildException {
        Map<String, String> sysProps = new HashMap<String, String>();

        String ivyCachePath = getIvyCachePath();
        sysProps.put("sbt.ivy.home", ivyCachePath);

        sysProps.putAll(JavaRunnerUtil.composeSystemProperties(getBuild(), getRunnerContext()));
        return sysProps;
    }

    @NotNull
    public String getClasspath() {
        String sbtHome = getSbtHome();
        File jarDir = new File(sbtHome, "bin");
        StringBuilder sb = new StringBuilder();
        for (String jar : SBT_JARS) {
            sb.append(new File(jarDir, jar).getAbsolutePath()).append(File.pathSeparator);
        }
        return sb.toString();
    }


    @NotNull
    private String getSbtHome() {
        String userSbtFolder = getRunnerParameters().get(SbtRunnerConstants.SBT_HOME_PARAM);
        if (isAutoInstallMode()) {
            return getAutoInstallSbtFolder();
        }
        return userSbtFolder;
    }

    private boolean isAutoInstallMode() {
        String sbtInstallationMode = getRunnerParameters().get(SbtRunnerConstants.SBT_INSTALLATION_MODE_PARAM);
        return SbtRunnerConstants.AUTO_INSTALL_FLAG.equalsIgnoreCase(sbtInstallationMode);
    }

    @NotNull
    public String getIvyCachePath() {
        return myIvyCacheProvider.getCacheDir().getAbsolutePath();
    }

    @NotNull
    public List<String> getCommands(SBTVersion sbtVersion) {
        String args = getRunnerParameters().get(SbtRunnerConstants.SBT_ARGS_PARAM).trim();
        if (StringUtil.isEmpty(args)) {
            getLogger().warning("No commands specified.");
            return Collections.emptyList();
        }
        return getCommandsFromFile(args, sbtVersion);
    }

    @NotNull
    private List<String> getCommandsFromFile(@NotNull String args, SBTVersion sbtVersion) {
        try {
            File file = FileUtil.createTempFile(getAgentTempDirectory(), "commands", ".file", true);
            String content = String.format(INFILE_COMMANDS_FORMATTER, getApplyCommand(sbtVersion), getCheckStatusCommand(), prepareArgs(args));
            String name = file.getAbsolutePath();
            getLogger().activityStarted("Prepare SBT run", "Write commands to file.", BUILD_ACTIVITY_TYPE);
            getLogger().message("File name: " + name + "; content: " + content);
            getLogger().activityFinished("Prepare SBT run", BUILD_ACTIVITY_TYPE);
            FileUtil.writeFile(file, content, "UTF-8");
            List<String> commands = new ArrayList<String>();
            String fileNameQuotes = SystemInfo.isWindows ? "\"\"" : "\"";
            commands.add(String.format(RUN_INFILE_COMMANDS_FORMATTER, fileNameQuotes + name.replace('\\', '/') + fileNameQuotes));
            return commands;
        } catch (IOException e) {
            LOG.warn(e.getMessage(), e);
            return Collections.emptyList();
        }
    }

    @NotNull
    private String prepareArgs(@NotNull String args) {
        if (args.startsWith(";")) {
            return args;
        }
        StringBuilder argss = new StringBuilder();
        String[] split = args.split(" ");
        for (String s : split) {
            String s1 = s.trim();
            if (!s1.isEmpty()) {
                argss.append("\n; ").append(s1);
            }
        }
        return argss.toString();
    }

    @Override
    public void afterProcessFinished() throws RunBuildException {
        super.afterProcessFinished();
        if ("false".equalsIgnoreCase(getConfigParameters().get("teamcity.internal.sbt.tempFilesCleanup.enabled"))) {
            return;
        }
        for (File file : myFilesToDelete) {
            FileUtil.delete(file);
        }
        myFilesToDelete.clear();
    }

    /**
     * SBT uses either XDG_RUNTIME_DIR env variable or java.io.tmpdir system property to determine its tmp dir.
     * java.io.tmpdir is equal to the buildTmp and XDG_RUNTIME_DIR is usually empty, so buildTmp is used by default.
     * We use XDG_RUNTIME_DIR here to override buildTmp in some cases.
     * ---
     * Why it's needed?
     * SBT stores "sbt-load.sock" socket in its tmp dir.
     * Unix domain sockets have a limitation that the path must be a maximum 108 characters.
     * The socket has the following path: "<tmpDirPath>/.sbt/sbt-socket<19 digits hash>/sbt-load.sock"
     * so the max allowed length of the <tmpDirPath> (buildTmp by default) can be 108 - 49 = 59.
     * ---
     * This workaround is needed after the upgrade of sbt-launch.jar from 1.5.5 to 1.10.10
     */
    private void fixSbtSocketLengthIssueIfNeeded(Map<String, String> envVars) throws RunBuildException {
        if ("false".equalsIgnoreCase(getConfigParameters().get("teamcity.internal.sbt.setXdgRuntimeDir.enabled"))) {
            return;
        }
        final int maxAllowedTmpDirLength = 59;
        if (SystemInfo.isWindows || getBuildTempDirectory().getAbsolutePath().length() <= maxAllowedTmpDirLength) {
            return;
        }
        final String xdgRuntimeDir = "XDG_RUNTIME_DIR";
        if (getEnvironmentVariables().containsKey(xdgRuntimeDir)) { // allow to override the value
            return;
        }
        final String sbtTmpDirPath = "/tmp/teamcity-sbt-" + RandomStringUtils.randomAlphanumeric(6);
        final File sbtTmpDir = new File(sbtTmpDirPath);
        try {
            FileUtil.createDir(sbtTmpDir);
        } catch (IOException e) {
            throw new RunBuildException("Failed to create temp directory for SBT at " + sbtTmpDirPath, e);
        }
        myFilesToDelete.add(sbtTmpDir);
        getLogger().message(xdgRuntimeDir + " set to: " + sbtTmpDirPath);
        envVars.put(xdgRuntimeDir, sbtTmpDirPath);
    }
}

File: tc-sbt-runner-agent/src/main/kotlin/jetbrains/buildServer/sbt/SbtRunnerBuildService.kt
package jetbrains.buildServer.sbt

import jetbrains.buildServer.RunBuildException
import jetbrains.buildServer.agent.runner.BuildServiceAdapter
import jetbrains.buildServer.agent.runner.CannotBuildCommandLineException
import jetbrains.buildServer.agent.runner.JavaCommandLineBuilder
import jetbrains.buildServer.agent.runner.JavaRunnerUtil
import jetbrains.buildServer.agent.runner.ProcessListener
import jetbrains.buildServer.agent.runner.ProcessListenerAdapter
import jetbrains.buildServer.agent.runner.ProgramCommandLine
import jetbrains.buildServer.messages.ErrorData
import jetbrains.buildServer.runner.JavaRunnerConstants
import jetbrains.buildServer.util.FileUtil
import jetbrains.buildServer.util.StringUtil
import org.apache.commons.lang3.RandomStringUtils
import org.apache.log4j.Logger
import java.io.File
import java.io.IOException
import java.util.jar.JarFile
import java.util.regex.Pattern

class SbtRunnerBuildService(
    private val ivyCacheProvider: IvyCacheProvider,
) : BuildServiceAdapter() {
    private val filesToDelete = mutableListOf<File>()

    enum class SBTVersion {
        SBT_1_x,
        SBT_0_13_x,
    }

    override fun getListeners(): List<ProcessListener> = listOf(
        object : ProcessListenerAdapter() {
            private var knownSectionStarts = false

            override fun onStandardOutput(line: String) {
                if (!knownSectionStarts && KNOWN_SECTION_MESSAGE.matcher(line).find()) {
                    knownSectionStarts = true
                }
                if (LINES_TO_EXCLUDE.matcher(line).find() && knownSectionStarts) {
                    return
                }
                logger.message(line)
            }

            override fun onErrorOutput(line: String) {
                logger.warning(line)
            }
        },
    )

    @Throws(RunBuildException::class)
    override fun makeProgramCommandLine(): ProgramCommandLine {
        val jvmArgs = JavaRunnerUtil.extractJvmArgs(runnerParameters)
        val mainClassName = if (isAutoInstallMode()) installSbt() else mainClassName
        val sbtVersion = SbtVersionDetector.discoverSbtVersion(workingDirectory, sbtLauncher, jvmArgs, logger)

        copySbtTcLogger(sbtVersion)

        val javaHome = javaHome
        val sbtHome = sbtHome
        logger.message("Java home set to: $javaHome")
        logger.message("SBT home set to: $sbtHome")

        val envVars = HashMap(environmentVariables)
        envVars[SbtRunnerConstants.SBT_HOME] = sbtHome
        envVars[JavaRunnerConstants.JAVA_HOME] = javaHome
        val path = envVars[PATH]
        envVars[PATH] = javaHome + if (!StringUtil.isEmpty(path)) File.pathSeparator + path else ""

        fixSbtSocketLengthIssueIfNeeded(envVars)

        val cliBuilder = JavaCommandLineBuilder()
        cliBuilder.setJavaHome(javaHome)
        cliBuilder.setBaseDir(checkoutDirectory.absolutePath)
        cliBuilder.setSystemProperties(vmProperties)
        cliBuilder.setEnvVariables(envVars)
        cliBuilder.setJvmArgs(jvmArgs)
        cliBuilder.setClassPath(classpath)
        cliBuilder.setMainClass(mainClassName)
        cliBuilder.setProgramArgs(getCommands(sbtVersion))
        cliBuilder.setWorkingDir(workingDirectory.absolutePath)

        return buildCommandline(cliBuilder)
    }

    private fun getApplyCommand(sbtVersion: SBTVersion): String {
        val pathToPlugin = File(
            autoInstallSbtFolder +
                File.separator +
                getPatchFolder(sbtVersion) +
                File.separator +
                SBT_PATCH_JAR_NAME,
        ).absolutePath
        return "apply -cp \"${pathToPlugin.replace('\\', '/')}\" $SBT_PATCH_CLASS_NAME"
    }

    private fun getCheckStatusCommand(): String = "sbt-teamcity-logger"

    @get:Throws(RunBuildException::class)
    private val javaHome: String
        get() {
            val home = JavaRunnerUtil.findJavaHome(
                runnerParameters[JavaRunnerConstants.TARGET_JDK_HOME],
                buildParameters.allParameters,
                checkoutDirectory.absolutePath,
            ) ?: throw RunBuildException("Unable to find Java home")
            return FileUtil.getCanonicalFile(File(home)).path
        }

    private val autoInstallSbtFolder: String
        get() = agentTempDirectory.toString() + File.separator + SBT_AUTO_HOME_FOLDER

    private fun copyResources(sourcePathInJar: String, sourceName: String, destinationDir: File) {
        destinationDir.mkdirs()
        val destination = File(destinationDir, sourceName)
        FileUtil.copyResource(javaClass, sourcePathInJar + sourceName, destination)
        if (destination.exists() && destination.isFile) {
            logger.message("Resource was copied to: $destination")
        }
    }

    private fun installSbt(): String =
        try {
            logger.activityStarted(
                SBT_INSTALLATION_STEP_NAME,
                "'Auto' mode was selected in SBT runner plugin settings",
                BUILD_ACTIVITY_TYPE,
            )
            logger.message("SBT will be install to: $autoInstallSbtFolder")
            copyResources(
                "/$SBT_DISTRIB/",
                SBT_LAUNCHER_JAR_NAME,
                File(autoInstallSbtFolder + File.separator + "bin"),
            )
            logger.message("SBT home set to: $sbtHome")
            mainClassName
        } catch (e: Exception) {
            logger.internalError(ErrorData.PREPARATION_FAILURE_TYPE, "An error occurred during SBT installation", e)
            throw IllegalStateException(e)
        } finally {
            logger.activityFinished(SBT_INSTALLATION_STEP_NAME, BUILD_ACTIVITY_TYPE)
        }

    private fun copySbtTcLogger(sbtVersion: SBTVersion) {
        try {
            logger.activityStarted(SBT_TEAMCITY_LOGGER_INSTALLATION, BUILD_ACTIVITY_TYPE)
            val to = autoInstallSbtFolder + File.separator + getPatchFolder(sbtVersion)
            val from = "/$SBT_DISTRIB/" + if (sbtVersion == SBTVersion.SBT_1_x) "$SBT_1_0_PATCH_FOLDER_NAME/" else ""
            logger.message(String.format("SBT logger %s will be installed from %s to %s", SBT_PATCH_JAR_NAME, from, to))
            copyResources(from, SBT_PATCH_JAR_NAME, File(to))
        } catch (e: Exception) {
            logger.internalError(ErrorData.PREPARATION_FAILURE_TYPE, "An error occurred during SBT installation", e)
            throw IllegalStateException(e)
        } finally {
            logger.activityFinished(SBT_TEAMCITY_LOGGER_INSTALLATION, BUILD_ACTIVITY_TYPE)
        }
    }

    private fun getPatchFolder(sbtVersion: SBTVersion): String =
        if (sbtVersion == SBTVersion.SBT_1_x) {
            SBT_PATCH_FOLDER_NAME + File.separator + SBT_1_0_PATCH_FOLDER_NAME
        } else {
            SBT_PATCH_FOLDER_NAME
        }

    @get:Throws(RunBuildException::class)
    private val mainClassName: String
        get() =
            try {
                val launcher = sbtLauncher
                logger.message("SBT main class name will be retrieved from: $launcher")
                JarFile(launcher).use { jarFile ->
                    jarFile.manifest.mainAttributes.getValue("Main-Class")
                        ?: throw RunBuildException("SBT launcher manifest does not define Main-Class")
                }
            } catch (e: IOException) {
                throw RunBuildException("An error occurred during reading manifest in SBT launcher", e)
            }

    private val sbtLauncher: File
        get() = File(File(sbtHome, "bin"), SBT_LAUNCHER_JAR_NAME)

    @Throws(RunBuildException::class)
    private fun buildCommandline(cliBuilder: JavaCommandLineBuilder): ProgramCommandLine =
        try {
            cliBuilder.build()
        } catch (e: CannotBuildCommandLineException) {
            throw RunBuildException(e.message ?: "Cannot build command line")
        }

    @get:Throws(RunBuildException::class)
    private val vmProperties: Map<String, String>
        get() {
            val sysProps = HashMap<String, String>()
            sysProps["sbt.ivy.home"] = ivyCachePath
            sysProps.putAll(JavaRunnerUtil.composeSystemProperties(build, runnerContext))
            return sysProps
        }

    val classpath: String
        get() {
            val jarDir = File(sbtHome, "bin")
            return SBT_JARS.joinToString(separator = File.pathSeparator, postfix = File.pathSeparator) { jar ->
                File(jarDir, jar).absolutePath
            }
        }

    private val sbtHome: String
        get() = if (isAutoInstallMode()) {
            autoInstallSbtFolder
        } else {
            runnerParameters[SbtRunnerConstants.SBT_HOME_PARAM].orEmpty()
        }

    private fun isAutoInstallMode(): Boolean =
        SbtRunnerConstants.AUTO_INSTALL_FLAG.equals(
            runnerParameters[SbtRunnerConstants.SBT_INSTALLATION_MODE_PARAM],
            ignoreCase = true,
        )

    val ivyCachePath: String
        get() = ivyCacheProvider.cacheDir.absolutePath

    fun getCommands(sbtVersion: SBTVersion): List<String> {
        val sbtCommands = runnerParameters[SbtRunnerConstants.SBT_ARGS_PARAM]?.trim().orEmpty()
        if (StringUtil.isEmpty(sbtCommands)) {
            logger.warning("No commands specified.")
            return emptyList()
        }
        return getCommandsFromFile(sbtCommands, sbtVersion)
    }

    private fun getCommandsFromFile(sbtCommands: String, sbtVersion: SBTVersion): List<String> =
        try {
            val file = FileUtil.createTempFile(agentTempDirectory, "commands", ".file", true)
            val prologueCommands = listOf(
                getApplyCommand(sbtVersion),
                getCheckStatusCommand(),
            )
            val content = SbtCommandFileContentBuilder.build(
                sbtCommands,
                prologueCommands,
            )
            val name = file.absolutePath
            logger.activityStarted("Prepare SBT run", "Write commands to file.", BUILD_ACTIVITY_TYPE)
            logger.message("File name: $name; content: $content")
            logger.activityFinished("Prepare SBT run", BUILD_ACTIVITY_TYPE)
            FileUtil.writeFile(file, content, "UTF-8")
            val fileNameQuotes = if (File.separatorChar == '\\') "\"\"" else "\""
            listOf(String.format(RUN_INFILE_COMMANDS_FORMATTER, fileNameQuotes + name.replace('\\', '/') + fileNameQuotes))
        } catch (e: IOException) {
            LOG.warn(e.message, e)
            emptyList()
        }

    override fun afterProcessFinished() {
        super.afterProcessFinished()
        if ("false".equals(configParameters["teamcity.internal.sbt.tempFilesCleanup.enabled"], ignoreCase = true)) {
            return
        }
        filesToDelete.forEach(FileUtil::delete)
        filesToDelete.clear()
    }

    /**
     * SBT uses either XDG_RUNTIME_DIR env variable or java.io.tmpdir system property to determine its tmp dir.
     * java.io.tmpdir is equal to buildTmp and XDG_RUNTIME_DIR is usually empty, so buildTmp is used by default.
     * We use XDG_RUNTIME_DIR here to override buildTmp in some cases.
     *
     * This workaround is needed after the upgrade of sbt-launch.jar from 1.5.5 to 1.10.10.
     */
    @Throws(RunBuildException::class)
    private fun fixSbtSocketLengthIssueIfNeeded(envVars: MutableMap<String, String>) {
        if ("false".equals(configParameters["teamcity.internal.sbt.setXdgRuntimeDir.enabled"], ignoreCase = true)) {
            return
        }
        if (environmentVariables.containsKey(XDG_RUNTIME_DIR)) {
            return
        }
        if (File.separatorChar == '\\' || buildTempDirectory.absolutePath.length <= MAX_ALLOWED_TMP_DIR_LENGTH) {
            return
        }

        val subDir = RandomStringUtils.randomAlphanumeric(4).lowercase()
        val tmpDirEnv = firstNonEmptyEnvVariableOrNull("TMPDIR", "TEMP", "TMP")
        val tmpDirRoot = if (tmpDirEnv != null && tmpDirEnv.length <= MAX_ALLOWED_TMP_DIR_LENGTH - subDir.length) {
            tmpDirEnv.removeSuffix("/")
        } else {
            "/tmp"
        }
        val tmpDir = File(tmpDirRoot, subDir)
        try {
            FileUtil.createDir(tmpDir)
        } catch (e: IOException) {
            throw RunBuildException("Failed to create temp directory for SBT at ${tmpDir.absolutePath}", e)
        }
        filesToDelete.add(tmpDir)
        logger.message("$XDG_RUNTIME_DIR set to: ${tmpDir.absolutePath}")
        envVars[XDG_RUNTIME_DIR] = tmpDir.absolutePath
    }

    private fun firstNonEmptyEnvVariableOrNull(vararg envVars: String): String? =
        envVars.asSequence()
            .mapNotNull(System::getenv)
            .firstOrNull(String::isNotBlank)

    companion object {
        private val LOG = Logger.getLogger(SbtRunnerBuildServiceFactory::class.java.name)

        private const val SBT_LAUNCHER_JAR_NAME = "sbt-launch.jar"
        private const val SBT_PATCH_JAR_NAME = "sbt-teamcity-logger.jar"
        private const val SBT_PATCH_FOLDER_NAME = "tc_plugin"
        private const val SBT_1_0_PATCH_FOLDER_NAME = "1.0"
        private const val SBT_DISTRIB = "sbt-distrib"
        private const val SBT_AUTO_HOME_FOLDER = "agent-sbt"
        private const val RUN_INFILE_COMMANDS_FORMATTER = "< %s"
        private const val SBT_INSTALLATION_STEP_NAME = "SBT installation"
        private const val SBT_TEAMCITY_LOGGER_INSTALLATION = "SBT TeamCity logger installation"
        private const val PATH = "PATH"
        private const val XDG_RUNTIME_DIR = "XDG_RUNTIME_DIR"
        private const val MAX_ALLOWED_TMP_DIR_LENGTH = 54
        private val SBT_JARS = arrayOf(SBT_LAUNCHER_JAR_NAME, "classes")

        const val BUILD_ACTIVITY_TYPE = "BUILD_ACTIVITY_TYPE"

        @JvmField
        val LINES_TO_EXCLUDE: Pattern = Pattern.compile(
            "^\\[(error|warn)\\]",
            Pattern.CASE_INSENSITIVE + Pattern.MULTILINE,
        )

        @JvmField
        val KNOWN_SECTION_MESSAGE: Pattern = Pattern.compile(
            "^(##teamcity\\[compilationStarted|testSuiteStarted|testStarted)",
            Pattern.CASE_INSENSITIVE + Pattern.MULTILINE,
        )

        private const val SBT_PATCH_CLASS_NAME = "jetbrains.buildServer.sbtlogger.SbtTeamCityLogger"

        @JvmStatic
        fun prepareArgs(args: String): String = SbtCommandFileContentBuilder.prepareArgs(args)
    }
}
